src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp

```diff
@@ -0,0 +1,1359 @@
+        "Check that " + from_expr_using_mode(ns, language_mode, lhs) +
+        " is assignable");
+
+      return false;
+    }
+
+    return true;
+  }
+  else
+  {
+    // This is a more complex expression.
+    // Since non-dirty locals are not tracked explicitly in the write set,
+    // we need to skip the check if we can verify that the expression describes
+    // an access to a non-dirty local symbol or an function parameter,
+    // otherwise the check will necessarily fail.
+    // If the expression contains address_of operator,
+    // the assignment gets checked. If the base object of the expression
+    // is a local or a function parameter, it will also be flagged as dirty and
+    // will be tracked explicitly, and the check will pass.
+    // If the expression contains a dereference operation, the assignment gets
+    // checked. If the dereferenced address was computed from a local object,
+    // from a function parameter or returned by a local malloc,
+    // then the object will also be tracked explicitly for being dirty,
+    // and the check will pass.
+    // In all other cases (address of a non-local object, or dereference of
+    // a non-locally computed address) the location must be given explicitly
+    // in the assigns clause to be allowed for assignment  and we must check
+    // the assignment.
+    if(cfg_info.is_local_composite_access(lhs))
+    {
+      check_source_location.set_comment(
+        "Check that " + from_expr_using_mode(ns, language_mode, lhs) +
+        " is assignable");
+
+      log.debug()
+        << "dfcc_instrumentt: Assignment to local symbol composite member "
+        << format(lhs) << messaget::eom;
+      return false;
+    }
+
+    log.debug() << "dfcc_instrumentt: checking assignment to expression "
+                << format(lhs) << messaget::eom;
+    return true;
+  }
+}
+
+void dfcc_instrumentt::instrument_lhs(
+  const irep_idt &function_id,
+  const exprt &write_set,
+  goto_programt::targett &target,
+  const exprt &lhs,
+  goto_programt &goto_program,
+  cfg_infot &cfg_info)
+{
+  const auto &mode = utils.get_function_symbol(function_id).mode;
+
+  goto_programt payload;
+
+  const auto &lhs_source_location = target->source_location();
+  auto goto_instruction = payload.add(goto_programt::make_incomplete_goto(
+    utils.make_null_check_expr(write_set), lhs_source_location));
+
+  source_locationt check_source_location(target->source_location());
+  check_source_location.set_property_class("assigns");
+
+  if(must_check_lhs(
+       target->source_location(), check_source_location, mode, lhs, cfg_info))
+  {
+    // ```
+    // IF !write_set GOTO skip_target;
+    // DECL check_assign: bool;
+    // CALL check_assign = check_assignment(write_set, &lhs, sizeof(lhs));
+    // ASSERT(check_assign);
+    // DEAD check_assign;
+    // skip_target: SKIP;
+    // ----
+    // ASSIGN lhs := rhs;
+    // ```
+
+    auto &check_sym = get_fresh_aux_symbol( // TODO use dfcc_utilst class
```

Comment:
About the TODO: You are using `get_fresh_aux_symbol` a lot, why should exactly this one be done differently?